### PR TITLE
feat: rtk swapper queries

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

Adds `swapperApi` with `getUsdRate` and `getTradeQuote` endpoints.

- `getUsdRate` partially implemented
- `getTradeQuote` not yet used due to current swapper architecture

Note, the `getUsdRate` `lib` method is still called a lot of times (so the network tab is only slightly better) as its currently being used as a hacky way to check for supported swapper pairs. It's gross, but PR does not intend to change that logic.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

Medium - touches swapper rate queries.

## Testing

Use the trade widget and check rates when selecting sell asset, buy asset, and using a fiat amount.

## Screenshots (if applicable)

N/A

### CODE REVIEW

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 1]
```diff
@@ -422,21 +405,17 @@ export const useSwapper = () => {
...
+        sellAssetUsdRate,
+        buyAssetUsdRate,
+        feeAssetUsdRate,
       }: DebouncedQuoteInput) => {
         try {
-          const [sellAssetUsdRate, buyAssetUsdRate, feeAssetUsdRate] = await Promise.all([
-            swapper.getUsdRate({ ...sellAsset }),
-            swapper.getUsdRate({ ...buyAsset }),
-            swapper.getUsdRate({ ...feeAsset }),
-          ])
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:438]: This removal is the meat of this PR. This logic is now handled by RTK Query at the beginning of `useSwapper`.

File: src/test/mocks/store.ts [hunk 2]
```diff
@@ -54,7 +55,6 @@ export const mockStore: ReduxState = {
       Avalanche: false,
       Thor: false,
       CowSwap: false,
-      JunoPay: false,
```
Comment (0xApotheosis) [inline src/test/mocks/store.ts:57]: `JunoPay` is not a key of `FeatureFlags`, I'm not sure why it was here.

**Review by 0xApotheosis — COMMENTED** [review]
Comments for reviewers.

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 3]
```diff
@@ -81,107 +75,100 @@ type DebouncedQuoteInput = {
...
+    sellAssetId: sellTradeAssetId,
+  })
+
+  const feeAssetFiatRateResponse = useGetUsdRateQuery({
+    rateAssetId: feeAsset?.assetId,
+    buyAssetId: buyTradeAssetId,
+    sellAssetId: sellTradeAssetId,
+  })
+
+  useEffect(() => {
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:159]: A bit of a strange pattern, I'm not convinced this is idiomatic.

File: src/state/apis/swapper/swapperApi.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+          return { data }
+        } catch (e) {
+          const data = `getAssetDescription: error fetching usd rate for ${rateAssetId}`
+          const status = 400
+          const error = { data, status }
+          return { error }
+        }
+      },
+    }),
+    getTradeQuote: build.query<TradeQuote<ChainId>, GetTradeQuoteInput>({
```
Comment (gomesalexandre) [inline src/state/apis/swapper/swapperApi.ts:68]: Would still abstract this to its own type for consistency
```suggestion
    getTradeQuote: build.query<GetTradeQuoteOutput, GetTradeQuoteInput>({
```

**Review by gomesalexandre — APPROVED** [review]
A bunch of smallish comments WRT query flow, types and error-handling but overall looks good so early stamping this so I can re-stamp this later.

Tested with ZRX and CoW swaps and didn't notice any regression, performance also seems slightly better from a user standpoint. Note that I didn't test Thor swaps as enabling the Thor flag currently rugs the trade input, but this is also happening in develop.

File: src/state/apis/swapper/swapperApi.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import { AssetId, ChainId } from '@shapeshiftoss/caip'
+import { GetTradeQuoteInput, Swapper, TradeQuote } from '@shapeshiftoss/swapper'
+import { getSwapperManager } from 'components/Trade/hooks/useSwapper/swapperManager'
+import { AssetsState } from 'state/slices/assetsSlice/assetsSlice'
+import { FeatureFlags, Preferences } from 'state/slices/preferencesSlice/preferencesSlice'
+
+type GetUsdRateArgs = {
+  rateAssetId: AssetId | undefined
+  buyAssetId: AssetId | undefined
+  sellAssetId: AssetId | undefined
```
[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/state/apis/swapper/swapperApi.ts:11]: yeah - we really can't avoid the blacklisting thing, we've got a ton of data in the store. 

bringing back logging middleware with some custom serialization is the best way to dev against this, it's very chatty but guaranteed to be correct.

**Comment by elmutt** [comment]
Looks legit. We could probably further improve swapper performance by looking at caching optimizations within the swappers themselves but this is great

File: src/components/Trade/hooks/useSwapper/swapperManager.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import { FeatureFlags } from 'state/slices/preferencesSlice/preferencesSlice'
+// import { store } from 'state/store'
+
+// singleton - do not export me, use getSwapperManager
+let _swapperManager: SwapperManager | null = null
+// singleton - do not export me
+// Used to short circuit calls to getSwapperManager if flags have not changed
+let previousFlags: string = ''
+
+export const getSwapperManager = async (flags: FeatureFlags): Promise<SwapperManager> => {
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Trade/hooks/useSwapper/swapperManager.ts:22]: For sure, though doing so causes some circular dependencies that are difficult to resolve.

Example in https://github.com/shapeshift/web/pull/2526/commits/2332bd0e531f09c88a7564630dc61c9b5aff8546, which was reverted.

[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Trade/hooks/useSwapper/swapperManager.ts:22]: I'll fix this up in the new swapper 👌

File: src/state/apis/swapper/swapperApi.ts (hunk 4 above)
Comment (0xApotheosis) [reply src/state/apis/swapper/swapperApi.ts]: For sure - fixed up in https://github.com/shapeshift/web/pull/2526/commits/00a8dc2ce49c18d6b93905b234cf52fef91ff491.

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 7]
```diff
@@ -81,107 +75,100 @@ type DebouncedQuoteInput = {
...
   }) as [
     TradeQuote<KnownChainIds> & Trade<KnownChainIds>,
     TradeAsset | undefined,
+    TradeAsset | undefined,
     Trade<KnownChainIds>,
     TradeState<KnownChainIds>['sellAssetAccount'],
     boolean,
+    string,
+    string,
+    string,
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts]: 💯 - https://github.com/shapeshift/web/pull/2526/commits/00a8dc2ce49c18d6b93905b234cf52fef91ff491

File: src/state/apis/swapper/swapperApi.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        try {
+          if (!rateAssetId) throw new Error('rateAssetId is undefined')
+          const swapper = await getBestSwapperFromArgs(buyAssetId, sellAssetId, featureFlags)
+          const rateAsset = assets.byId[rateAssetId]
+          const usdRate = await swapper.getUsdRate(rateAsset)
+          const data = { usdRate }
+          return { data }
+        } catch (e) {
+          const data = `getAssetDescription: error fetching usd rate for ${rateAssetId}`
+          const status = 400
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/apis/swapper/swapperApi.ts]: Good callout, returning a `400` would be inaccurate.

Updated in https://github.com/shapeshift/web/pull/2526/commits/69bf6fd08314e3fa415740283412a858f150d15a.

File: src/state/apis/swapper/swapperApi.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+          const swapper = await getBestSwapperFromArgs(
+            args.sellAsset.assetId,
+            args.buyAsset.assetId,
+            featureFlags,
+          )
+          const tradeQuote = await swapper.getTradeQuote(args)
+          return { data: tradeQuote }
+        } catch (e) {
+          const data = 'getTradeQuote: error fetching trade quote'
+          const status = 400
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/apis/swapper/swapperApi.ts]: I've used a `CUSTOM_ERROR ` for now in https://github.com/shapeshift/web/pull/2526/commits/69bf6fd08314e3fa415740283412a858f150d15a - you're right, we should come up with and implement a best-practice way to handle error handling in our queries.

File: src/state/apis/swapper/swapperApi.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/apis/swapper/swapperApi.ts]: Using skip is a much nicer pattern, nice find @gomesalexandre 🤙 

Used in https://github.com/shapeshift/web/pull/2526/commits/faa48b1a39298e33dae2d6e89b66ee0cdeb11d42.

**Comment by 0xApotheosis** [comment]
This is good to go once Cypress unrugs itself and CI goes green.

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**